### PR TITLE
810: Session vs non-session routes

### LINKED ISSUE #810: Split Express routing into session vs non-session
In order to prevent unnecessary sessions being created when we hit e.g. the health check endpoint (which the load balancer does regularly) we neede to re-configure the express middleware and routes so that the session is only used on endpoints that need to read or write session data.

This work was originally done on the cloud infrastructure branch, but I'd like it in dev first, so that the cloud branch _only_ has infra code.

### PR DESCRIPTION
**Pull request by chriswilty**
## Description

We need to split API routes into session and non-session endpoints, so that we are not generating or noop-updating the session when we don't need to. This is particularly important for cloud deployment, where our load-balancer hits our healthcheck every 30secs, creating a new session every time 😱 

Resolves #810 

## Checklist

Have you done the following?

- [x] Linked the relevant Issue
- [ ] Added tests
- [x] Ensured the workflow steps are passing

### CODE REVIEW

File: backend/src/app.ts [hunk 1]
```diff
@@ -1,74 +1,16 @@
...
-import { importMetaUrl } from './importMetaUtils';
-import { ChatModel, defaultChatModel } from './models/chat';
-import { LevelState, getInitialLevelStates } from './models/level';
-import { router } from './router';
+import nonSessionRoutes from './nonSessionRoutes';
+import sessionRoutes from './sessionRoutes';
 
-dotenv.config();
-
-declare module 'express-session' {
```
Comment (chriswilty) [inline backend/src/app.ts:15]: Much of this file has been moved to sessionRoutes.ts, the rest in nonSessionRoutes.ts

File: backend/src/nonSessionRoutes.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+router.use(
+	'/documents',
+	express.static(
+		fileURLToPath(new URL('../resources/documents', importMetaUrl()))
+	)
+);
+router.get('/documents', handleGetDocuments);
+router.get('/health', handleHealthCheck);
+router.get('/systemRoles', handleGetSystemRoles);
```
Comment (chriswilty) [inline backend/src/nonSessionRoutes.ts:19]: There might be one or two more endpoints that don't need a session, e.g. validModels, but these will do for now.

File: backend/package.json [hunk 3]
```diff
@@ -23,8 +23,7 @@
 		"memorystore": "^1.6.7",
 		"openai": "^4.19.0",
 		"openai-chat-tokens": "^0.2.8",
-		"pdf-parse": "^1.1.1",
-		"react": "^18.2.0"
```
Comment (pmarsh-scottlogic) [inline backend/package.json:27]: good spot lol

File: backend/src/langchain.ts [hunk 4]
```diff
@@ -20,11 +20,12 @@ import {
...
 // store vectorised documents for each level as array
 const vectorisedDocuments = (() => {
-	let docs: DocumentsVector[] = [];
+	const docs: DocumentsVector[] = [];
 	return {
 		get: () => docs,
 		set: (newDocs: DocumentsVector[]) => {
-			docs = newDocs;
+			while (docs.length > 0) docs.pop();
+			docs.push(...newDocs);
```
Comment (pmarsh-scottlogic) [inline backend/src/langchain.ts:28]: Is there a particular reason why you did this? Just curious, not upset with it.

File: backend/src/models/chat.ts [hunk 5]
```diff
@@ -141,20 +141,15 @@ export type {
 	ChatDefenceReport,
 	ChatGptReply,
 	ChatMalicious,
+	ChatModel,
+	ChatModelConfiguration,
```
Comment (pmarsh-scottlogic) [inline backend/src/models/chat.ts:145]: another good spot 👍

File: backend/src/sessionRoutes.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+if (!sessionSigningSecret) {
+	console.error(
+		'SESSION_SECRET not found in environment vars, cannot continue!'
+	);
+	process.exit(1);
+}
+
+const router = express.Router();
+
+const stage = process.env.NODE_ENV;
```
Comment (pmarsh-scottlogic) [inline backend/src/sessionRoutes.ts:51]: stage seems like a weird name, but then I'm not familiar with the terminology.

**Review by pmarsh-scottlogic — CHANGES_REQUESTED** [review]
Struggling to run this locally. Some (but not all) endpoints come back with CORS errors

loading up the frontend

image

Sending a chat message

image

testing on postman the endpoints seem fine.

File: backend/src/langchain.ts (hunk 4 above)
Comment (chriswilty) [reply backend/src/langchain.ts:28]: There was - I figured probably better to use a const within the closure; we can then be certain we are not holding onto someone else's copy of the document array, and therefore our copy will not change without us knowing about it.

With the original code, we were replacing our reference entirely, so fully at the mercy of any other code with access to that reference. Of course, the contents are only shallow-copied so the actual items could be mutated, but hopefully we are avoiding as much mutation as possible now in our codebase.

File: backend/src/sessionRoutes.ts (hunk 6 above)
Comment (chriswilty) [reply backend/src/sessionRoutes.ts:51]: Yeah it's an AWS thing 🤐 

They use "environment" to mean "region and account", whereas what we know as environments are called stages. Which is doubly odd when it's quite common to have a "staging" environment, so then you'd have "staging" stage 😕

**Comment by chriswilty** [comment]
@pmarsh-scottlogic Ah yes sorry, you'll need to replace your backend `.env` with the new `.env.example` as I added the cors origin into there.

File: backend/src/langchain.ts (hunk 4 above)
Comment (pmarsh-scottlogic) [reply backend/src/langchain.ts:28]: right, makes sense

File: backend/src/sessionRoutes.ts (hunk 6 above)
Comment (pmarsh-scottlogic) [reply backend/src/sessionRoutes.ts:51]: Ah well! Can't win them all

**Review by pmarsh-scottlogic — APPROVED** [review]
Works a charm

File: backend/test/integration/langchain.test.ts [hunk 7]
```diff
@@ -49,16 +51,15 @@ jest.mock('langchain/embeddings/openai', () => {
 	};
 });
 
-class MockMemoryVectorStore {
-	asRetriever() {
-		mockAsRetriever();
-	}
-}
```
Comment (chriswilty) [inline backend/test/integration/langchain.test.ts:56]: @pmarsh-scottlogic It seemed unnecessary to create a class, so I inlined it below.

File: backend/src/document.ts [hunk 8]
```diff
@@ -28,6 +28,7 @@ async function getDocuments(filePath: string) {
 		'.csv': (path: string) => new CSVLoader(path),
 	});
 	const docs = await loader.load();
+	console.debug(`${docs.length} documents found`);
```
Comment (chriswilty) [inline backend/src/document.ts:31]: This is the line that exposed we weren't stubbing the loader to return anything. So I simply added that in my test changes.

File: backend/test/integration/langchain.test.ts [hunk 9]
```diff
@@ -115,28 +116,28 @@ jest.mock('@src/openai', () => {
...
-	mockRetrievalQAChain.call.mockRestore();
-	mockFromLLM.mockRestore();
-	mockFromTemplate.mockRestore();
-});
+	afterEach(() => {
+		mockPromptEvalChain.call.mockReset();
+		mockRetrievalQAChain.call.mockReset();
+		mockFromLLM.mockReset();
+		mockFromTemplate.mockReset();
+		mockLoader.mockReset();
```
Comment (pmarsh-scottlogic) [inline backend/test/integration/langchain.test.ts:135]: could you use `mockClear()` instead and then not have to do `mockLoader.mockResolvedValue([]);` in the beforeEach?

[doc](mockLoader.mockResolvedValue([]);)

Comment (chriswilty) [reply backend/test/integration/langchain.test.ts]: Done

**Review by pmarsh-scottlogic — APPROVED** [review]
lovely

**Pull request merged**